common/dynamicconfig/collection.go

```diff
@@ -0,0 +1,543 @@
+	// Available value types:
+	//   Bool: bool
+	//   Duration: time.Duration
+	//   Float: float64
+	//   Int: int
+	//   Map: map[string]any
+	//   String: string
+	// Available filters:
+	//   Namespace func(namespace string)
+	//   NamespaceID func(namespaceID string)
+	//   TaskQueueInfo func(namespace string, taskQueue string, taskType enumspb.TaskQueueType)
+	//   ShardID func(shardID int32)
+	BoolPropertyFn                             func() bool
+	BoolPropertyFnWithNamespaceFilter          func(namespace string) bool
+	BoolPropertyFnWithNamespaceIDFilter        func(namespaceID string) bool
+	BoolPropertyFnWithTaskQueueInfoFilters     func(namespace string, taskQueue string, taskType enumspb.TaskQueueType) bool
+	DurationPropertyFn                         func() time.Duration
+	DurationPropertyFnWithNamespaceFilter      func(namespace string) time.Duration
+	DurationPropertyFnWithNamespaceIDFilter    func(namespaceID string) time.Duration
+	DurationPropertyFnWithShardIDFilter        func(shardID int32) time.Duration
+	DurationPropertyFnWithTaskQueueInfoFilters func(namespace string, taskQueue string, taskType enumspb.TaskQueueType) time.Duration
+	FloatPropertyFn                            func() float64
+	FloatPropertyFnWithNamespaceFilter         func(namespace string) float64
+	FloatPropertyFnWithShardIDFilter           func(shardID int32) float64
+	FloatPropertyFnWithTaskQueueInfoFilters    func(namespace string, taskQueue string, taskType enumspb.TaskQueueType) float64
+	IntPropertyFn                              func() int
+	IntPropertyFnWithNamespaceFilter           func(namespace string) int
+	IntPropertyFnWithShardIDFilter             func(shardID int32) int
+	IntPropertyFnWithTaskQueueInfoFilters      func(namespace string, taskQueue string, taskType enumspb.TaskQueueType) int
+	MapPropertyFn                              func() map[string]any
+	MapPropertyFnWithNamespaceFilter           func(namespace string) map[string]any
+	StringPropertyFn                           func() string
+	StringPropertyFnWithNamespaceFilter        func(namespace string) string
+)
+
+const (
+	errCountLogThreshold = 1000
+)
+
+var (
+	errKeyNotPresent        = errors.New("key not present")
+	errNoMatchingConstraint = errors.New("no matching constraint in key")
+)
+
+// NewCollection creates a new collection
+func NewCollection(source Source, logger log.Logger) *Collection {
+	return &Collection{
+		source:   source,
+		logger:   logger,
+		errCount: -1,
+	}
+}
+
+func (c *Collection) logError(key Key, err error) {
+	// TODO: This is a lot of unnecessary contention with little benefit. Consider using
+	// https://github.com/cespare/percpu here.
+	errCount := atomic.AddInt64(&c.errCount, 1)
+	if errCount%errCountLogThreshold == 0 {
+		// log only every 'x' errors to reduce mem allocs and to avoid log noise
+		c.logger.Debug("No such key in dynamic config, using default", tag.Key(key.String()), tag.Error(err))
```

Comment:
it could be key not found, or it could be errNoMatchingConstraint, right?

Reply:
yeah. I just left the message as is. the error is in the tag.

also note that for partition count, since the structured defaults count as constraints, it'll say errNoMatchingConstraint even when the key is missing and it should really say errKeyNotPresent. I could fix that with a little more logic, but again, not sure it's worth it.